### PR TITLE
Implement ability to mount parts of the worker filesystem. Fixes #5495

### PR DESCRIPTION
**Pull request by twitchyliquid64**
## What does this PR accomplish?

Bug Fix | **Feature** | Documentation

closes #5495, #5674.

## Changes proposed by this PR:
As discussed in #5495, a new option is added to ATC to apply additional bind mounts to all task steps. This should allow people who _do_ trust their workloads with access to the workers, to do so. 

This is necessary for access to attached peripherals (such as GPUs or hardware testbeds), and then also kids like me who run concourse on a single box and just want to yeet artifacts to disk.

## Notes to reviewer:
Basically adding a `--mount-holepunch "workerpath:containerpath"` should create the corresponding bind mount in all task step executions.

Reading the checklist below ... what do

## Contributor Checklist

- [X] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [X] [Signed] all commits
- [X] Added tests (**Unit** and/or Integration)
- [ ] Updated [Documentation]
- [ ] Updated [Release notes]

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs
[Release notes]: https://github.com/concourse/concourse/tree/master/release-notes

## Reviewer Checklist

- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

### CODE REVIEW

File: atc/atccmd/command.go [hunk 1]
```diff
@@ -243,6 +244,10 @@ type RunCommand struct {
 	EnableArchivePipeline bool     `long:"enable-archive-pipeline" description:"Enable /api/v1/teams/{team}/pipelines/{pipeline}/archive endpoint."`
 
 	EnableBuildRerunWhenWorkerDisappears bool `long:"enable-rerun-when-worker-disappears" description:"Enable automatically build rerun when worker disappears"`
+
+	Unsafe struct {
+		MountHolepunches map[string]string `long:"mount-holepunch" description:"Binds a path on all workers into each task step. Can be specified multiple times." value-name:"FROM-PATH:TO-PATH"`
+	} `group:"Unsafe options (optional)"`
```
Comment (vito) [inline atc/atccmd/command.go:250]: Nice, I think this definitely makes it suitably scary. +1 for holepunch terminology.

File: atc/engine/builder/builder.go [hunk 2]
```diff
@@ -21,7 +22,7 @@ const supportedSchema = "exec.v2"
 type StepFactory interface {
 	GetStep(atc.Plan, exec.StepMetadata, db.ContainerMetadata, exec.GetDelegate) exec.Step
 	PutStep(atc.Plan, exec.StepMetadata, db.ContainerMetadata, exec.PutDelegate) exec.Step
-	TaskStep(atc.Plan, exec.StepMetadata, db.ContainerMetadata, exec.TaskDelegate) exec.Step
+	TaskStep(atc.Plan, exec.StepMetadata, db.ContainerMetadata, exec.TaskDelegate, atc.UnsafeWorkerOverrides) exec.Step
```
Comment (vito) [inline atc/engine/builder/builder.go:25]: Good idea to not pass this along to the resource containers. That'd be a whole can of worms.

File: atc/atccmd/command.go [hunk 3]
```diff
@@ -1447,6 +1458,18 @@ func (cmd *RunCommand) constructLockConn(driverName string) (*sql.DB, error) {
 	return dbConn, nil
 }
 
+func (cmd *RunCommand) parseUnsafeOverrides() (atc.UnsafeWorkerOverrides, error) {
+	for fromPath, toPath := range cmd.Unsafe.MountHolepunches {
+		if !filepath.IsAbs(fromPath) {
+			return atc.UnsafeWorkerOverrides{}, fmt.Errorf("mount-holepunch source must be an an absolute path")
+		}
+		if !filepath.IsAbs(toPath) {
```
Comment (vito) [inline atc/atccmd/command.go:1466]: 👍 to this being absolute - I think it technically *could* be relative and effectively mount in to each task's working directory, but I could see that causing conflicts with regular build inputs

File: atc/engine/builder/builder.go (hunk 2 above)
Comment (vito) [reply atc/engine/builder/builder.go:25]: Makes sense!

**Comment by maxnbk** [comment]
I made a pass at addressing merge conflicts in https://github.com/maxnbk/concourse/tree/issue/5495 ,
but can't merge it to this PR since its @twitchyliquid64  's.
(I also can't address @xtreme-sameer-vohra  's comments, as I'm just a python dev trying to be able to use concourse in our facility, but can't proceed without this feature, or one like it, to be able to)
If I can buy someone a weeks coffee to polish/finish this, I'll do it.

**Comment by twitchyliquid64** [comment]
I'm happy to merge your commits into this PR if you want?

**Comment by maxnbk** [comment]
By all means, @twitchyliquid64 (if I did it correctly)

**Comment by twitchyliquid64** [comment]
Aaaaaaah what happened

**Comment by twitchyliquid64** [comment]
Gonna try fixing this

**Comment by twitchyliquid64** [comment]
@vito something autoclosed the PR but I can't re-open it, are you able?

**Comment by zoetian** [comment]
maybe this will help? @twitchyliquid64 
https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c

**Comment by twitchyliquid64** [comment]
Okay so its not gonna merge ... started working on fixing the changes, but they are too great and I've already had two drinks tonight.

Im committing what I have because there are some things that can be re-used if someone wants to give re-implementing this a go.

**Comment by muntac** [comment]
Closing this as it's only meant to serve as inspiration for a re-implementation at this point ude to the merge conflicts.

**Pull request closed without merging**